packages/enzyme/src/selectors.js

```diff
@@ -0,0 +1,318 @@
+import { createParser } from 'rst-selector-parser';
+import isEmpty from 'lodash/isEmpty';
+import unique from 'lodash/uniq';
+import {
+  treeFilter,
+  nodeHasId,
+  findParentNode,
+  nodeMatchesObjectProps,
+  childrenOfNode,
+  hasClassName,
+} from './RSTTraversal';
+import { nodeHasType, nodeHasProperty } from './Utils';
+// our CSS selector parser instance
+const parser = createParser();
+
+// Combinators that allow you to chance selectors
+const CHILD = 'childCombinator';
+const ADJACENT_SIBLING = 'adjacentSiblingCombinator';
+const GENERAL_SIBLING = 'generalSiblingCombinator';
+const DESCENDANT = 'descendantCombinator';
+
+// Selectors for targeting elements
+const SELECTOR = 'selector';
+const TYPE_SELECTOR = 'typeSelector';
+const CLASS_SELECTOR = 'classSelector';
+const ID_SELECTOR = 'idSelector';
+const ATTRIBUTE_PRESENCE = 'attributePresenceSelector';
+const ATTRIBUTE_VALUE = 'attributeValueSelector';
+// @TODO we dont support these, throw if they are used
+const PSEUDO_CLASS = 'pseudoClassSelector';
+const PSEUDO_ELEMENT = 'pseudoElementSelector';
+
+/**
+ * Calls reduce on a array of nodes with the passed 
+ * function, returning only unique results.
+ * @param {Function} fn 
+ * @param {Array<Node>} nodes 
+ */
+function uniqueReduce(fn, nodes) {
+  return unique(nodes.reduce(fn, []));
+}
+
+/**
+ * Takes a CSS selector and returns a set of tokens parsed
+ * by scalpel.
+ * @param {String} selector 
+ */
+function safelyGenerateTokens(selector) {
+  try {
+    return parser.parse(selector);
+  } catch (err) {
+    throw new Error(`Failed to parse selector: ${selector}`);
+  }
+}
+
+/**
+ * Takes a node and a token and determines if the node
+ * matches the predicate defined by the token.
+ * @param {Node} node 
+ * @param {Token} token 
+ */
+function nodeMatchesToken(node, token) {
+  if (node === null || typeof node === 'string') {
```

Comment:
can you elaborate on why a string node wouldn't match? is there no selector that can match text?

Reply:
Correct, with the current grammar there's no way to match a string literal. `foo` would match an element of type `foo`. We *could* augment the grammar to allow something like `"foo"` but I'm not sure how useful that really is.